### PR TITLE
refactor : Transform SaveModal to typescript

### PR DESCRIPTION
**Pull request by maloun96**
### SUMMARY
Transform SaveModal to Typescript

### CODE REVIEW

**Comment by maloun96** [comment]
@evans @junlincc

File: superset-frontend/src/dashboard/components/SaveModal.tsx [hunk 1]
```diff
@@ -27,33 +26,49 @@ import ModalTrigger from '../../components/ModalTrigger';
...
-  colorNamespace: PropTypes.string,
-  colorScheme: PropTypes.string,
-  onSave: PropTypes.func.isRequired,
-  canOverwrite: PropTypes.bool.isRequired,
-  refreshFrequency: PropTypes.number.isRequired,
-  lastModifiedTime: PropTypes.number.isRequired,
+type SaveType = typeof SAVE_TYPE_OVERWRITE | typeof SAVE_TYPE_NEWDASHBOARD;
+
+type SaveModalProps = {
+  addSuccessToast: Function;
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/SaveModal.tsx:32]: Instead of `Function` can you define the arguments and return values expected in the function here too?`

File: superset-frontend/src/dashboard/components/SaveModal.tsx [hunk 2]
```diff
@@ -27,33 +26,49 @@ import ModalTrigger from '../../components/ModalTrigger';
...
+  addDangerToast: Function;
+  dashboardId: number;
+  dashboardTitle: string;
+  dashboardInfo: object;
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/SaveModal.tsx:36]: Prefer `Record<string, any>` to `object`

File: superset-frontend/src/dashboard/components/SaveModal.tsx [hunk 3]
```diff
@@ -69,25 +84,25 @@ class SaveModal extends React.PureComponent {
     this.onSave = this.props.onSave.bind(this);
   }
 
-  setModalRef(ref) {
+  setModalRef(ref: ModalTrigger | null) {
     this.modal = ref;
   }
 
-  toggleDuplicateSlices() {
+  toggleDuplicateSlices(): any {
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/SaveModal.tsx:91]: Why does this return `any`? seems like nothing is returned

File: superset-frontend/src/dashboard/components/SaveModal.tsx [hunk 4]
```diff
@@ -116,7 +131,7 @@ class SaveModal extends React.PureComponent {
     // check refresh frequency is for current session or persist
     const refreshFrequency = shouldPersistRefreshFrequency
       ? currentRefreshFrequency
-      : dashboardInfo.metadata.refresh_frequency; // eslint-disable camelcase
+      : (dashboardInfo as any).metadata.refresh_frequency; // eslint-disable camelcase
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/SaveModal.tsx:134]: This is a bit worrisome, can we do this without the `any`?

File: superset-frontend/src/dashboard/components/SaveModal.tsx [hunk 5]
```diff
@@ -137,17 +152,17 @@ class SaveModal extends React.PureComponent {
         t('You must pick a name for the new dashboard'),
       );
     } else {
-      this.onSave(data, dashboardId, saveType).then(resp => {
+      this.onSave(data, dashboardId, saveType).then((resp: any) => {
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/SaveModal.tsx:155]: I believe a `SupersetResponse` type exists, perhaps in `superset-ui/core`? That would be preferable to `any` here

File: superset-frontend/src/dashboard/components/SaveModal.tsx [hunk 6]
```diff
@@ -137,17 +152,17 @@ class SaveModal extends React.PureComponent {
...
         if (
           saveType === SAVE_TYPE_NEWDASHBOARD &&
           resp &&
           resp.json &&
           resp.json.id
         ) {
-          window.location = `/superset/dashboard/${resp.json.id}/`;
+          window.location.href = `/superset/dashboard/${resp.json.id}/`;
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/SaveModal.tsx:162]: oo, was this a bug that typescript caught?

**Review by etr2460 — CHANGES_REQUESTED** [review]

File: superset-frontend/src/dashboard/components/SaveModal.tsx [hunk 7]
```diff
@@ -69,25 +87,25 @@ class SaveModal extends React.PureComponent {
     this.onSave = this.props.onSave.bind(this);
   }
 
-  setModalRef(ref) {
+  setModalRef(ref: ModalTrigger | null) {
     this.modal = ref;
   }
 
-  toggleDuplicateSlices() {
+  toggleDuplicateSlices(val?: boolean | undefined): void {
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/SaveModal.tsx:94]: you don't need the `| undefined` if you already have a `?:`

Although, I wonder why you added an argument here since it seems like the function doesn't actually use it? Is that because you pass it into a component that expects a function with this arg?

File: superset-frontend/src/dashboard/components/SaveModal.tsx [hunk 8]
```diff
@@ -18,42 +18,60 @@
...
+
+type SaveType = typeof SAVE_TYPE_OVERWRITE | typeof SAVE_TYPE_NEWDASHBOARD;
+
+type SaveModalProps = {
+  addSuccessToast: (arg: string) => void;
+  addDangerToast: (arg: string) => void;
+  dashboardId: number;
+  dashboardTitle: string;
+  dashboardInfo: Record<string, any>;
+  expandedSlices: object;
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/SaveModal.tsx:40]: Could you update all instances of `object`?

File: superset-frontend/src/dashboard/components/SaveModal.tsx [hunk 9]
```diff
@@ -18,42 +18,60 @@
...
 };
 
-class SaveModal extends React.PureComponent {
-  constructor(props) {
+class SaveModal extends React.PureComponent<SaveModalProps, SaveModalState> {
+  static defaultProps = defaultProps;
+
+  modal: ModalTrigger | null;
+
+  onSave: Function;
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/SaveModal.tsx:72]: and all instances of `Function`? Thanks!

File: superset-frontend/src/components/Checkbox/index.tsx [hunk 10]
```diff
@@ -25,8 +25,8 @@ import {
 
 interface CheckboxProps {
   checked: boolean;
-  onChange: (val?: boolean) => {};
-  style: object;
+  onChange: (val?: boolean) => void;
+  style?: object;
```
Comment (etr2460) [inline superset-frontend/src/components/Checkbox/index.tsx:29]: are these changes intentional? I don't remember them before

**Comment by maloun96** [comment]
@etr2460 Refactored.

File: superset-frontend/src/components/Checkbox/index.tsx (hunk 10 above)
Comment (rusackas) [reply superset-frontend/src/components/Checkbox/index.tsx:29]: _If_ it _is_ intentional (I'm not sure either), it'd be the only remaining instance of `object` so maybe we can touch that up too?

Comment (ktmud) [reply superset-frontend/src/components/Checkbox/index.tsx:29]: Shouldn't it be `import { CSSProperties } from 'react'`?

File: superset-frontend/src/dashboard/components/SaveModal.tsx (hunk 6 above)
Comment (ktmud) [reply superset-frontend/src/dashboard/components/SaveModal.tsx:169]: Setting `location` is equivalent to setting `location.href`.

Comment (etr2460) [reply superset-frontend/src/dashboard/components/SaveModal.tsx:169]: Wow, you calling me a youngin?!? 😆

File: superset-frontend/src/dashboard/components/SaveModal.tsx [hunk 11]
```diff
@@ -185,7 +207,7 @@ class SaveModal extends React.PureComponent {
             <div className="m-l-25 m-t-5">
               <Checkbox
                 checked={this.state.duplicateSlices}
-                onChange={this.toggleDuplicateSlices}
+                onChange={() => this.toggleDuplicateSlices}
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/SaveModal.tsx:210]: wondering why this was needed, seems like it might break it (not that toggleDuplicateSlices isn't actually being called I think)

**Review by etr2460 — COMMENTED** [review]
One more comment from me (and also @rusackas/@ktmud's comment too)

File: superset-frontend/src/dashboard/components/SaveModal.tsx (hunk 6 above)
Comment (rusackas) [reply superset-frontend/src/dashboard/components/SaveModal.tsx:169]: Well, _back when I_ started building websites, there wasn't even JavaScript yet! Or CSS! Gather round, let me spin you a yarn about the old days. You see, there was this browser called Mosaic...

... hey, are you kids even listening?

**Review by rusackas — APPROVED** [review]
LGTM! Thanks all for the comments, the fixes, and the reminiscing about the olden days.

File: superset-frontend/src/dashboard/components/SaveModal.tsx [hunk 12]
```diff
@@ -185,7 +207,7 @@ class SaveModal extends React.PureComponent {
             <div className="m-l-25 m-t-5">
               <Checkbox
                 checked={this.state.duplicateSlices}
-                onChange={this.toggleDuplicateSlices}
+                onChange={() => this.toggleDuplicateSlices()}
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/SaveModal.tsx:210]: still a bit confused why we needed to change this since we bind this in the constructor for this function

**Review by etr2460 — APPROVED** [review]
lgtm with one final question

File: superset-frontend/src/dashboard/components/SaveModal.tsx [hunk 13]
```diff
@@ -18,42 +18,64 @@
...
+  saveType: SaveType;
+  newDashName: string;
+  duplicateSlices: boolean;
 };
 
 const defaultProps = {
   saveType: SAVE_TYPE_OVERWRITE,
   colorNamespace: undefined,
   colorScheme: undefined,
+  shouldPersistRefreshFrequency: true,
```
Comment (graceguo-supercat) [inline superset-frontend/src/dashboard/components/SaveModal.tsx:64]: why `shouldPersistRefreshFrequency` default be true?

Comment (etr2460) [reply superset-frontend/src/dashboard/components/SaveModal.tsx:64]: oh, good catch

File: superset-frontend/src/dashboard/components/SaveModal.tsx (hunk 12 above)
Comment (maloun96) [reply superset-frontend/src/dashboard/components/SaveModal.tsx:210]: We get an warning of that argumeng "boolean | undefined" where we do not use it. That's why I added callback here

File: superset-frontend/src/dashboard/components/SaveModal.tsx (hunk 13 above)
Comment (maloun96) [reply superset-frontend/src/dashboard/components/SaveModal.tsx:64]: changed back to false!

**Comment by rusackas** [comment]
I think all questions/issues have been resolved! Thanks, everyone!
